File: noisecluster-jvm/src/test/scala/noisecluster/test/specs/transport/aeron/BasicSpec.scala
/**
  * Copyright 2017 https://github.com/sndnv
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
package noisecluster.test.specs.transport.aeron

import akka.actor.ActorSystem
import io.aeron.Aeron
import io.aeron.driver.MediaDriver
import noisecluster.test.utils._
import noisecluster.transport.aeron.{Defaults, Source, Target}
import org.scalatest.{FlatSpec, Matchers}

import scala.concurrent.{ExecutionContext, Future}
import scala.util.Random

class BasicSpec extends FlatSpec with Matchers {
  private var testDataSent = 0L
  private var testDataReceived = 0L

  private def testDataHandler: (Array[Byte], Int) => Unit = (_: Array[Byte], length: Int) => {
    testDataReceived += length
  }

  private implicit val ec: ExecutionContext = ExecutionContext.Implicits.global
  private implicit val loggingSystem = ActorSystem("testLoggingSystem")
  private val channel = "aeron:ipc"
  private val stream = 42
  private val driver = MediaDriver.launch(Defaults.getNewDriverContext)
  private implicit val aeron = Aeron.connect(Defaults.getNewSystemContext)

  private val source: Source = Source(stream, channel, Defaults.BufferSize)
  private val target: Target = Target(stream, channel, testDataHandler, Defaults.IdleStrategy, Defaults.FragmentLimit)

  private val testByteArraySize = 1000

  private var targetFuture = Future {
    target.start()
  }

  waitUntil(what = "target becomes inactive", waitTimeMs = 500, waitAttempts = 10) {
    target.isActive
  }

  "A source and a target" should "successfully exchange data" in {
    val bytes = Array.ofDim[Byte](testByteArraySize)
    Random.nextBytes(bytes)
    source.send(bytes)
    testDataSent += testByteArraySize

    waitUntil(what = "data is received by target", waitTimeMs = 500, waitAttempts = 10) {
      testDataSent == testDataReceived
    }

    testDataSent should be(testDataReceived)
    testDataReceived should be(testByteArraySize)
  }

  "A target" should "successfully stop accepting data" in {
    target.isActive should be(true)
    target.stop()

    waitUntil(what = "target becomes inactive", waitTimeMs = 500, waitAttempts = 10) {
      !target.isActive
    }

    target.isActive should be(false)

    assertThrows[IllegalStateException] {
      target.stop()
    }
  }

  it should "successfully restart and accept data" in {
    target.isActive should be(false)
    targetFuture = Future {
      target.start()
    }

    waitUntil(what = "target becomes active", waitTimeMs = 500, waitAttempts = 10) {
      target.isActive
    }

    target.isActive should be(true)

    val bytes = Array.ofDim[Byte](testByteArraySize)
    Random.nextBytes(bytes)
    source.send(bytes)
    testDataSent += testByteArraySize

    waitUntil(what = "data is received by target", waitTimeMs = 500, waitAttempts = 10) {
      testDataSent == testDataReceived
    }

    testDataSent should be(testDataReceived)
    testDataReceived should be(testByteArraySize * 2)
  }

  it should "fail to close an active connection" in {
    assertThrows[IllegalStateException] {
      target.close()
    }
  }

  it should "successfully stop and close its connection" in {
    target.isActive should be(true)
    target.stop()

    waitUntil(what = "target becomes inactive", waitTimeMs = 500, waitAttempts = 10) {
      !target.isActive
    }

    target.close()
    target.isActive should be(false)
  }

  "A system" should "successfully terminate" in {
    aeron.close()
    driver.close()
  }
}

File: noisecluster-win/noisecluster/win/test/transport/aeron/BasicSpec.cs
/**
  * Copyright 2017 https://github.com/sndnv
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

using System;
using System.Threading.Tasks;
using Adaptive.Aeron;
using log4net.Config;
using noisecluster.win.transport;
using noisecluster.win.transport.aeron;
using NUnit.Framework;

namespace noisecluster.win.test.transport.aeron
{
    [TestFixture]
    public class BasicSpec
    {
        private long _testDataSent;
        private long _testDataReceived;

        private readonly DataHandler _testDataHandler;

        private readonly Aeron _aeron;

        private readonly AeronSource _source;
        private readonly AeronTarget _target;

        private readonly int _testByteArraySize;

        private readonly Random _rnd;
        private Task _targetTask;

        public BasicSpec()
        {
            BasicConfigurator.Configure();
            _testDataSent = 0;
            _testDataReceived = 0;

            _testDataHandler = (data, length) => { _testDataReceived += length; };

            const string channel = "aeron:ipc";
            const int stream = 42;
            _aeron = Aeron.Connect(Defaults.GetNewSystemContext());

            _source = new AeronSource(_aeron, stream, channel, Defaults.BufferSize);
            _target = new AeronTarget(_aeron, stream, channel, Defaults.IdleStrategy, Defaults.FragmentLimit);

            _testByteArraySize = 1000;

            _rnd = new Random();

            _targetTask = new Task(() => { _target.Start(_testDataHandler); });
            _targetTask.Start();
        }

        [Test]
        public void T01_SourceAndTarget_should_ExchangeData()
        {
            var bytes = new byte[_testByteArraySize];
            _rnd.NextBytes(bytes);
            _source.Send(bytes);
            _testDataSent += _testByteArraySize;

            Utils.WaitUntil("data is received by target", 500, 10, () => _testDataSent == _testDataReceived);

            Assert.IsTrue(_testDataSent == _testDataReceived);
            Assert.IsTrue(_testDataReceived == _testByteArraySize);
        }

        [Test]
        public void T02_Target_should_StopAcceptingData()
        {
            Assert.IsTrue(_target.IsActive);
            _target.Stop();

            Utils.WaitUntil("target becomes inactive", 500, 10, () => !_target.IsActive);

            Assert.IsFalse(_target.IsActive);

            Assert.Throws<InvalidOperationException>(() => { _target.Stop(); });
        }

        [Test]
        public void T03_Target_should_RestartAcceptingData()
        {
            Assert.IsFalse(_target.IsActive);

            _targetTask = new Task(() => { _target.Start(_testDataHandler); });
            _targetTask.Start();

            Utils.WaitUntil("target becomes active", 500, 10, () => _target.IsActive);

            Assert.IsTrue(_target.IsActive);

            var bytes = new byte[_testByteArraySize];
            _rnd.NextBytes(bytes);
            _source.Send(bytes);
            _testDataSent += _testByteArraySize;

            Utils.WaitUntil("data is received by target", 500, 10, () => _testDataSent == _testDataReceived);

            Assert.IsTrue(_testDataSent == _testDataReceived);
            Assert.IsTrue(_testDataReceived == _testByteArraySize * 2);
        }

        [Test]
        public void T04_Target_should_FailToCloseActiveConnection()
        {
            Assert.Throws<InvalidOperationException>(() => { _target.Close(); });
        }

        [Test]
        public void T05_Target_should_StopAndCloseItsConnection()
        {
            Assert.IsTrue(_target.IsActive);
            _target.Stop();

            Utils.WaitUntil("target becomes inactive", 500, 10, () => !_target.IsActive);

            _target.Close();
            Assert.IsFalse(_target.IsActive);
        }

        [Test]
        public void T06_System_should_DisposeOfTransportObjects()
        {
            _source.Dispose();
            _target.Dispose();
            _aeron.Dispose();
        }
    }
}